### ISSUE TITLE
no-member false negative for dynamic classes

### ISSUE THREAD

**Issue by LarsMichelsen** [PyCQA/pylint#2851]
### Steps to reproduce

1. no-member for self.b in last line is found as expected:

```
class A:
    def __init__(self):
        self.a = 1

class B(A):
    def __init__(self):
        print(self.a)
        print(self.b)
```

2. self.b access is not reported as issue:

```
if __debug__:
    class A:
        def __init__(self):
            self.aaa = 1
else:
    class A:
        def __init__(self):
            self.aaa = 1

class B(A):
    def __init__(self):
        print(self.a)
        print(self.b)
```

### Current behavior

Once a class is defined in ca conditional, like __debug__ here, pylint silently ignores skips the no-member test.

I have discovered this in reality for a Python 2.7 class that subclasses threading.Thread. It seems that the subclassing issue of threading.Thread was cleaned up between Python 2 and 3, but the issue with such situations in pylint remains.

### Expected behavior

Optimum would be to report the no-member issue in this situation, but I can imagine that this is not easily possible.

In such a situation I would expect pylint to at least issue some informational message about the fact that this test can not be executed on the classes in this class hierarchy.

If you look at the code of some checkers, it looks more like a fundamental question. My point is:  In the moment you use tools like pylint, you count on certain features. As soon as these, from the user's point of view, don't work as expected under certain circumstances, the tool's usefulness fades.

### pylint --version output
```

pylint 2.3.1
astroid 2.2.5
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0]
```

**Comment by PCManticore** [comment]
Thanks for the report @LarsMichelsen There might be something tripping the inference here, which leads to the error not being emitted.

**Comment by LarsMichelsen** [comment]
Did some quick digging into the code:

- checkers.typecheck.visit_attribute() goes to the exceptions.NotFoundError catching case: https://github.com/PyCQA/pylint/blob/c8db761f5fbe7535115639d578cd46c350d25b47/pylint/checkers/typecheck.py#L886
- It calls _emit_no_member() which returns False. This makes the no-member-Issue be skipped silently with the continue statement here: https://github.com/PyCQA/pylint/blob/c8db761f5fbe7535115639d578cd46c350d25b47/pylint/checkers/typecheck.py#L901
- In _emit_no_member() we have the problematic part in https://github.com/PyCQA/pylint/blob/c8db761f5fbe7535115639d578cd46c350d25b47/pylint/checkers/typecheck.py#L419

**Comment by mbyrnepr2** [comment]
On latest Pylint `no-member` is emitted as expected for `self.b` in the given examples 1 & 2. Perhaps this can be closed.

**Comment by DanielNoord** [comment]
Thanks @mbyrnepr2! I'm closing this :)
